README.org

```diff
@@ -371,19 +373,163 @@ We can also change the thickness of the lines and shapes that we draw by changin
+
+In this example, we draw a new rectangle every time there is a click. The important parameters of the event method: =state= is a keyword symbol indicating the state of the mouse, =ts= is a timestamp, =x= and =y= are the coordinates of the mouse click.
+
+#+BEGIN_SRC lisp
+  (defsketch input-test
+     ((title "Hello, input")
+      (rectangles nil))
+   (loop for (x y) in rectangles
+         do (rect x y 50 50)))
+  (defmethod kit.sdl2:mousebutton-event ((window input-test) state ts b x y)
+    (with-slots (rectangles) window
+      (when (eq state :mousebuttondown)
+        (push (list x y) rectangles))))
+#+END_SRC
+
+In this example, all keyboard text input is echoed to the screen. =text= is a string / character array containing a single character.
+
+#+BEGIN_SRC lisp
+  (defsketch input-test
+     ((title "Hello, input")
+      (text-to-write nil))
+    (loop for s in text-to-write
+          do (text s 0 0 20 20)
+          do (translate 20 0)))
+  (defmethod kit.sdl2:textinput-event ((window input-test) ts text)
+    (with-slots (text-to-write) window
+      (setf text-to-write (nconc text-to-write (list text)))))
+#+END_SRC
+
+Finally, here is an example where a pair of eyes follow the mouse (the pupils are restricted to a rectangle, it would look better if they were restricted to a circle).
+
+#+BEGIN_SRC lisp
+  (defsketch input-test
+     ((looking-at (list 0 0))
+      (cx (/ width 2))
+      (cy (/ height 2)))
+    (let ((cx-1 (- cx 50))
+          (cx-2 (+ cx 50))
+          (mx (car looking-at))
+          (my (cadr looking-at)))
+      (with-pen (make-pen :fill +white+)
+        (ellipse cx-1 cy 40 80)
+        (ellipse cx-2 cy 40 80))
+      (with-pen (make-pen :fill +black+)
+        (flet ((move-towards (x1 x2)
+                 (let ((diff (- x2 x1)))
+                   (+ x1 (if (< (abs diff) 10)
+                             diff
+                             (* (signum diff) 10))))))
+          (circle (move-towards cx-1 mx) (move-towards cy my) 10)
+          (circle (move-towards cx-2 mx) (move-towards cy my) 10)))))
+  (defmethod kit.sdl2:mousemotion-event ((window input-test) ts bm x y xrel yrel)
+    (with-slots (looking-at) window
+      (setf (car looking-at) x
+            (cadr looking-at) y)))
+#+END_SRC
+
+See also: [[https://github.com/vydd/sketch/blob/master/examples/life.lisp][life.lisp]] and [[https://github.com/vydd/sketch/blob/master/examples/input.lisp][input.lisp]].
+
+*** Setup
+The generic function =(setup instance &key &allow-other-keys)= is a hook that gets called before the window is initialised. Since the window has not been initialised, it can't be used to draw anything, but the background colour can be set. Here is an example from [[https://github.com/vydd/sketch/blob/master/examples/brownian.lisp][brownian.lisp]].
```

Comment:
Actually the doc string is in the wrong here - the `setup` method is called once on every "restart":
- before calling the `draw` for the first time;
- when the sketch is redefined
- every time an error occurs (after the red screen).

Reply:
Thus it also can draw anything it wants, but that will only work with `:copy-pixels t` because otherwise it will be immediately wiped by the `(background (gray 0.4))` call inside the internal wrapper around the draw method.